framework/src/loops/ComputeResidualAndJacobianThread.C

```diff
@@ -0,0 +1,173 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeResidualAndJacobianThread.h"
+#include "NonlinearSystem.h"
+#include "Problem.h"
+#include "FEProblem.h"
+#include "KernelBase.h"
+#include "IntegratedBCBase.h"
+#include "DGKernelBase.h"
+#include "InterfaceKernelBase.h"
+#include "Material.h"
+#include "TimeKernel.h"
+#include "SwapBackSentinel.h"
+#include "FVTimeKernel.h"
+
+#include "libmesh/threads.h"
+
+ComputeResidualAndJacobianThread::ComputeResidualAndJacobianThread(
```

Comment:
sorry - not trying to be rude.

this for example looks like it duplicates quite a lot from the standard residual/jacobian loop, and also misses quite a bit of functionality. with such duplication, it becomes easier and easier to forget about objects when adding features

Reply:
Some of the object forgetting can be purposeful 😄 But it is definitely possible to forget by mistake although I've tried to implement a good error message that would make it easy to know what is missing

There is a lot of duplication between this and `ComputeJacobianThread` and `ComputeResidualThread`. You can pretty easily remove some duplication in things like `subdomainChanged`. But it's harder for the `onElement`, `onBoundary`, `onInternalSide`, etc. methods because at the core they have to call `computeResidual` or `computeJacobian` or `computeResidualAndJacobian`. I don't think we want to go down the road of changing those user interfaces to something like `compute(Moose::ComputeType)`. Maybe the thing to do would be to do what I did in `ComputeFVFluxThread` which is create a generic loop with a `virtual compute(ResidualObject &)` that derived classes like `ComputeResidualThread` override to dispatch to `computeResidual()` via like
```c++
ComputeResidualThread::compute(ResidualObject & ro)
{
  ro.computeResidual();
}
```

Reply:
☹️ even for kernels, there are subtle differences
`ComputeResidualThread`:
```c++
  if (_tag_kernels->hasActiveBlockObjects(_subdomain, _tid))
  {
    const auto & kernels = _tag_kernels->getActiveBlockObjects(_subdomain, _tid);
    for (const auto & kernel : kernels)
      kernel->computeResidual();
  }
```
`ComputeJacobianThread`:
```c++
  if (_warehouse->hasActiveBlockObjects(_subdomain, _tid))
  {
    auto & kernels = _warehouse->getActiveBlockObjects(_subdomain, _tid);
    for (const auto & kernel : kernels)
      if (kernel->isImplicit())
      {
        kernel->prepareShapes(kernel->variable().number());
        kernel->computeJacobian();
	/// done only when nonlocal kernels exist in the system                                                                                                                                                                               
        if (_fe_problem.checkNonlocalCouplingRequirement())
        {
          std::shared_ptr<NonlocalKernel> nonlocal_kernel =
              std::dynamic_pointer_cast<NonlocalKernel>(kernel);
          if (nonlocal_kernel)
            kernel->computeNonlocalJacobian();
        }
      }
  }
```